File: src/main/java/ru/l0sty/frogdisplays/cef/CefUtil.java
package ru.l0sty.frogdisplays.cef;

import com.cinemamod.mcef.MCEF;
import com.cinemamod.mcef.MCEFBrowser;
import ru.l0sty.frogdisplays.CinemaModClient;
import ru.l0sty.frogdisplays.screen.Screen;
import org.cef.CefApp;
import org.cef.CefClient;
import org.cef.CefSettings;

public final class CefUtil {

    private CefUtil() {}

    private static boolean init;
    private static CefApp cefAppInstance;
    private static CefClient cefClientInstance;

    public static boolean init() {
//        String[] cefSwitches = new String[]{
//                "--autoplay-policy=no-user-gesture-required",
//                "--disable-web-security"
//        };
//
//        if (!CefApp.startup(cefSwitches)) {
//            return false;
//        }
//
//        CefSettings cefSettings = new CefSettings();
//        cefSettings.windowless_rendering_enabled = true;
//        cefSettings.background_color = cefSettings.new ColorType(0, 255, 255, 255);
//
//        cefAppInstance = CefApp.getInstance(cefSwitches, cefSettings);
//
////        CefApp.addAppHandler(new CefCinemaAppHandler(cefSwitches));
//
//        cefClientInstance = cefAppInstance.createClient();
//        cefClientInstance.addLoadHandler(new CefBrowserCinemaLoadHandler());

        return init = true;
    }

    public static boolean isInit() {
        return init;
    }

    public static CefApp getCefApp() {
        return cefAppInstance;
    }

    public static CefClient getCefClient() {
        return cefClientInstance;
    }

    public static MCEFBrowser createBrowser(String startUrl, int widthPx, int heightPx) {
        return MCEF.createBrowser(startUrl, false, widthPx, heightPx);
//        if (!init) return null;
//        CefBrowserCinema browser = new CefBrowserCinema(cefClientInstance, startUrl, false, null);
//        browser.setCloseAllowed();
//        browser.createImmediately();
//        browser.resize(widthPx, heightPx);
//        return browser;
    }

    public static MCEFBrowser createBrowser(String startUrl, Screen screen) {
        if (!init) return null;
        MCEFBrowser browser = MCEF.createBrowser(startUrl, true);
        browser.setCloseAllowed();
        browser.createImmediately();
        // Adjust screen size
        {
            float widthBlocks = screen.getWidth();
            float heightBlocks = screen.getHeight();
            float scale = widthBlocks / heightBlocks;
            int height = CinemaModClient.getInstance().getVideoSettings().getBrowserResolution();
            int width = (int) Math.floor(height * scale);
            browser.resize(width, height);
        }
        return browser;
    }

}

File: src-unused/client/kotlin/ru/l0sty/frogdisplays/cef/CefUtil.kt
package ru.l0sty.frogdisplays.cef

import org.cef.CefApp
import org.cef.CefClient
import org.cef.CefSettings
import kotlin.math.floor

object CefUtil {
    var isInit: Boolean = false
        private set
    var cefApp: CefApp? = null
        private set
    var cefClient: CefClient? = null
        private set

    fun init(): Boolean {
        val cefSwitches = arrayOf(
            "--autoplay-policy=no-user-gesture-required",
            "--disable-web-security"
        )

        if (!CefApp.startup(cefSwitches)) {
            return false
        }

        val cefSettings = CefSettings()
        cefSettings.windowless_rendering_enabled = true
        cefSettings.background_color = cefSettings.ColorType(0, 255, 255, 255)

        cefApp = CefApp.getInstance(cefSwitches, cefSettings)

        //        CefApp.addAppHandler(new CefCinemaAppHandler(cefSwitches));
        cefClient = cefApp!!.createClient()
        //cefClient.addLoadHandler(CefBrowserCinemaLoadHandler())

        return true.also { isInit = it }
    }

    fun createBrowser(startUrl: String?, widthPx: Int, heightPx: Int): CefBrowserCinema? {
        if (!isInit) return null
        val browser = CefBrowserCinema(cefClient, startUrl, false, null)
        browser.setCloseAllowed()
        browser.createImmediately()
        browser.resize(widthPx, heightPx)
        return browser
    }

    fun createBrowser(startUrl: String?): CefBrowserCinema? {
        if (!isInit) return null
        val browser = CefBrowserCinema(cefClient, startUrl, true, null)
        browser.setCloseAllowed()
        browser.createImmediately()
        // Adjust screen size
        run {
            val widthBlocks: Float = 1f
            val heightBlocks: Float = 1f
            val scale = widthBlocks / heightBlocks
            val height: Int = 400
            val width = floor((height * scale).toDouble()) as Int
            browser.resize(width, height)
        }
        return browser
    }
}